src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +238,331 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
+            {
+                addAndAdjustOldDelta(builder, ref oldDelta, oldChange);
+                tryGetNextOldChange();
+            }
+
+            while (newIndex < newChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
+                addAdjustedNewChange(builder, oldDelta, newChange);
+                tryGetNextNewChange();
+            }
+
+            return builder.ToImmutableAndFree();
+
+            bool tryGetNextOldChange()
+            {
+                oldIndex++;
+                if (oldIndex < oldChanges.Length)
+                {
+                    oldChange = oldChanges[oldIndex];
+                    return true;
+                }
+                else
                 {
-                    AddRange(list, oldChanges[oldIndex]);
-                    oldIndex++;
+                    oldChange = default;
+                    return false;
                 }
             }
 
-            return list.ToImmutableArray();
-        }
-
-        private static void AddRange(List<TextChangeRange> list, TextChangeRange range)
-        {
-            if (list.Count > 0)
+            bool tryGetNextNewChange()
             {
-                var last = list[list.Count - 1];
-                if (last.Span.End == range.Span.Start)
+                newIndex++;
+                if (newIndex < newChanges.Length)
                 {
-                    // merge changes together if they are adjacent
-                    list[list.Count - 1] = new TextChangeRange(new TextSpan(last.Span.Start, last.Span.Length + range.Span.Length), last.NewLength + range.NewLength);
-                    return;
+                    newChange = new UnadjustedNewChange(newChanges[newIndex]);
+                    return true;
                 }
                 else
                 {
-                    Debug.Assert(range.Span.Start > last.Span.End);
+                    newChange = default;
+                    return false;
                 }
             }
 
-            list.Add(range);
+            static void addAndAdjustOldDelta(ArrayBuilder<TextChangeRange> builder, ref int oldDelta, TextChangeRange oldChange)
+            {
+                // modify oldDelta to reflect characters deleted and inserted by an old change
+                oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
+                add(builder, oldChange);
+            }
+
+            static void addAdjustedNewChange(ArrayBuilder<TextChangeRange> builder, int oldDelta, UnadjustedNewChange newChange)
+            {
+                // unadjusted new change is relative to the original text with old changes applied. Subtract oldDelta to make it relative to the original text.
+                add(builder, new TextChangeRange(new TextSpan(newChange.SpanStart - oldDelta, newChange.SpanLength), newChange.NewLength));
```

Comment:
What about boundary casess. What if say the original doc was 2^31 characters long, the oldDelta is now -2^31 cause it deleted everything and we add it to any non-zero value in `newChange.SpanStart`. That will suddenly be outside the bounds of an `int` an overflow correct?

Reply:
If the old change deleted everything, then the only valid `newChange.SpanStart` position is 0. The overflow would result in a negative number, right? I think the failure mode would be TextSpan's constructor throwing due to being given a negative Start.